instrumentation/opentelemetry-instrumentation-aiohttp-client/src/opentelemetry/instrumentation/aiohttp_client/__init__.py

```diff
@@ -203,3 +241,84 @@ def _trace_config_ctx_factory(**kwargs):
     trace_config.on_request_exception.append(on_request_exception)
 
     return trace_config
+
+
+def _instrument(
+    tracer_provider: TracerProvider = None,
+    url_filter: _UrlFilterT = None,
+    span_name: _SpanNameT = None,
+):
+    """Enables tracing of all ClientSessions
+
+    When a ClientSession gets created a TraceConfig is automatically added to
+    the session's trace_configs.
+    """
+    # pylint:disable=unused-argument
+    def instrumented_init(wrapped, instance, args, kwargs):
+        if context_api.get_value("suppress_instrumentation"):
```

Comment:
We also have this check in instrumented methods. Does this exist to prevent instrumenting libraries in the first place or to prevent generating spans for certain code paths? I think it's the later. In other words, I think we should still instrument the libraries but instrumentation itself should check for this for each request and skip generating spans in case.

Reply:
once `AioHttpClientInstrumentor().instrument()` is called, the library is always instrumented. this check just supresses instrumentation for a single `ClientSession`.
`aiohttp` already comes with built in hooks for instrumenting certain aspects. To make use of these hooks one has to pass one or more `TraceConfig` objects to the `ClientSession` constructor. The `AioHttpClientInstrumentor` piggybacks on this mechanism by wrapping the `ClientSession` constructor to automatically add a `TraceConfig` with the respective hooks every time a `ClientSession` is created. The `suppress_instrumentation` is to exclude single `ClientSession`s from being instrumented.